micrometer-meter-provider/src/main/java/io/opentelemetry/contrib/metrics/micrometer/MicrometerMeterProvider.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.metrics.micrometer;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.opentelemetry.api.metrics.MeterBuilder;
+import io.opentelemetry.api.metrics.MeterProvider;
+import io.opentelemetry.contrib.metrics.micrometer.internal.state.MeterProviderSharedState;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * An implementation of {@link MeterProvider} that delegates metrics to a Micrometer {@link
+ * MeterRegistry}.
+ */
+public final class MicrometerMeterProvider implements MeterProvider, AutoCloseable {
+  static final String OTEL_POLLING_METER_NAME = "otel_polling_meter";
+
+  private final MeterProviderSharedState meterProviderSharedState;
+  private final CallbackRegistrar callbackRegistrar;
+
+  /**
+   * Creates a new instance of {@link MicrometerMeterProvider} for the provided {@link
+   * MeterRegistry}.
+   *
+   * @param meterRegistrySupplier supplies the {@link MeterRegistry}
+   */
+  MicrometerMeterProvider(
+      Supplier<MeterRegistry> meterRegistrySupplier, CallbackRegistrar callbackRegistrar) {
+    this.callbackRegistrar = callbackRegistrar;
+    this.meterProviderSharedState =
+        new MeterProviderSharedState(meterRegistrySupplier, callbackRegistrar);
+  }
+
+  /** Closes the current provider. */
+  @Override
+  public void close() {
+    callbackRegistrar.close();
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public MeterBuilder meterBuilder(String instrumentationScopeName) {
+    Objects.requireNonNull(instrumentationScopeName, "instrumentationScopeName");
+    return new MicrometerMeterBuilder(meterProviderSharedState, instrumentationScopeName);
+  }
+
+  /** Returns a new builder instance for this provider with the specified {@link MeterRegistry}. */
+  public static MicrometerMeterProviderBuilder builder(MeterRegistry meterRegistry) {
+    Objects.requireNonNull(meterRegistry, "meterRegistry");
+    return new MicrometerMeterProviderBuilder(() -> meterRegistry);
+  }
+
+  /**
+   * Returns a new builder instance for this provider with a {@link Supplier} for a {@link
+   * MeterRegistry}.
+   */
+  public static MicrometerMeterProviderBuilder builder(Supplier<MeterRegistry> meterRegistry) {
```

Comment:
I did this mostly for Spring as it has unfortunate behavior if you materialize the MeterRegistry bean too early during initialization, it results in the BeanPostProcessors which handle the MeterRegistryCustomizers getting skipped.  But you'd likely want to setup the MeterProvider during Spring bootstrap as well.

Reply:
I see. I'm a bit worried that someone will pass something like `MeterRegistry::new` that'll create a new instance every time it's called. WDYT about clearly documenting that it's supposed to be used as a late/lazy initialization mechanism, and wrapping the supplier in an equivalent of guava's `MemoizingSupplier` so that the original one is never called more than once?